ckan/logic/action/update.py

```diff
@@ -296,16 +326,54 @@ def package_update(
 
         resource_uploads.append(upload)
 
+    validate_data = dict(data_dict)
+    if changed_resources is not None:
+        validate_data['resources'] = changed_resources
+
     data, errors = lib_plugins.plugin_validate(
-        package_plugin, context, data_dict, schema, 'package_update')
+        package_plugin,
+        context,
+        validate_data,
+        schema,
+        'package_update'
+    )
+    # shift resource errors to the correct positions
+    if copy_resources and 'resources' in errors:
+        rerrs = []
+        i = 0
+        for e in errors['resources']:
+            while i in copy_resources:
+                rerrs.append({})
+                i += 1
+            rerrs.append(e)
+            i += 1
+        errors['resources'] = rerrs
     log.debug('package_update validate_errs=%r user=%s package=%s data=%r',
               errors, user, context['package'].name, data)
 
     if errors:
         model.Session.rollback()
         raise ValidationError(errors)
 
-    #avoid revisioning by updating directly
+    # merge skipped resources back into validated data
+    if copy_resources:
```

Comment:
Can you explain this snippet? I get that we want to merge the skipped resources back with the changed ones but I don't understand the second while loop

Reply:
the second while loop adds anything remaining in copy_resources after the end of `data['resources']` is reached